Handle <attachment> painting in GPUP

nmoucht opened the pull request:
#### b4421d4a471ca4a0d2c02b67713a81380c441725
<pre>
Handle &lt;attachment&gt; painting in GPUP
<a href="https://bugs.webkit.org/show_bug.cgi?id=246368">https://bugs.webkit.org/show_bug.cgi?id=246368</a>
rdar://83459796

Reviewed by Simon Fraser.

Unify some parts of layout/painting of attachment on mac and ios.
This includes the struct containing layout info of the attachment
element, the functions constructing the lines of the subtitle and
title, painting of the lines. Move the bits constructing the
layout struct out of RenderThemeMac/iOS into AttachmentLayout.

* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::paint):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::AttachmentLayout::layOutSubtitle):
(WebCore::paintAttachmentIcon):
(WebCore::drawCTLine):
(WebCore::paintAttachmentTitle):
(WebCore::paintAttachmentSubtitle):

Canonical link: <a href="https://commits.webkit.org/255903@main">https://commits.webkit.org/255903@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3935f9d2e92f71e09cda8509daaefabec4bdc5db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93952 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3141 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24517 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103585 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163922 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97945 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3157 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31380 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86274 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99640 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99615 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2260 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80375 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29276 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84186 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72232 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37769 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17710 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35639 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18975 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4075 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39512 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41543 "Found 1 new test failure: fast/workers/stress-js-execution.html (failure)") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41451 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38205 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/AttachmentLayout.h:
```diff
@@ -0,0 +1,80 @@
+struct AttachmentLayout {
+    explicit AttachmentLayout(const RenderAttachment&, AttachmentLayoutStyle style = AttachmentLayoutStyle::NonSelected);
+    
+    float widthPadding;
```
Add initializers for all the plain data values.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/AttachmentLayout.h:
```diff
@@ -0,0 +1,80 @@
+    CGFloat contentYOrigin { 0 };
+    void layOutSubtitle(const RenderAttachment&);
+    void layOutTitle(const RenderAttachment&);
+    void buildWrappedLines(String&, CTFontRef, NSDictionary*, unsigned);
```
Please name the arguments.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -458,12 +460,20 @@ bool RenderTheme::paint(const RenderBox& box, ControlStates& controlStates, cons
+    ControlPart part = box.style().effectiveAppearance();
+    IntRect integralSnappedRect = snappedIntRect(rect);
+
+    // Temporarily move this call above the canPaint check to allow
```
This says "Temporarily" but is there a long-term fix?

nmoucht replied:
Maybe a good solution to this is to explicitly allow painting of certain `ControlPart` in `canPaint` without a platform context.

smfr commented on LayoutTests/fast/attachment/mac/attachment-element-gpu-process.html:
```diff
@@ -0,0 +1,6 @@
+<!DOCTYPE html><!-- webkit-test-runner [ AttachmentElementEnabled=true UseGPUProcessForDOMRenderingEnabled=true ] -->
```
This test will stop testing anything meaningful when UseGPUProcessForDOMRenderingEnabled is turned on by default.

nmoucht replied:
I explicitly set `UseGPUProcessForDOMRenderingEnabled = false` in the expected html. This does beg the question of whether we still want to test the non gpup code path after this is enabled. Maybe it would make sense to remove this test once it is enabled by default.

smfr replied:
A better test would be to compare <attachment> rendering against some content that is not <attachment> but has similar rendering.

smfr commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1118,6 +1123,20 @@ bool RenderTheme::paintAttachment(const RenderObject&, const PaintInfo&, const I
 
+int RenderTheme::attachmentBaseline(const RenderAttachment& attachment) const
+{
+    AttachmentLayout layout(attachment, AttachmentLayoutStyle::NonSelected);
```
Do we end up running the layout code multiple times? Maybe that's not new, but it would nice to avoid.

nmoucht replied:
Yes this is how it worked before. It could be useful to try and figure out how to only recompute the attachment layout when needed.

smfr commented on Source/WebCore/rendering/AttachmentLayout.mm:
```diff
@@ -0,0 +1,397 @@
+    if (subtitleText.isEmpty())
+        return;
+    Color subtitleColor = attachment.style().colorByApplyingColorFilter(attachmentSubtitleTextColor);
+    CFStringRef language = 0; // By not specifying a language we use the system language.
```
nullptr

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/AttachmentLayout.mm:
```diff
@@ -0,0 +1,397 @@
+    auto& subtitleText = attachment.attachmentElement().attributeWithoutSynchronization(HTMLNames::subtitleAttr);
+    if (subtitleText.isEmpty())
+        return;
+    Color subtitleColor = attachment.style().colorByApplyingColorFilter(attachmentSubtitleTextColor);
```
auto

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/AttachmentLayout.mm:
```diff
@@ -0,0 +1,397 @@
+    if (text.isEmpty())
+        return;
+    
+    RetainPtr<NSAttributedString> attributedText = adoptNS([[NSAttributedString alloc] initWithString:text attributes:textAttributes]);
```
Could use lots more `auto` in this function.

nmoucht replied:
Fixed.

smfr reviewed (approved):
r=me but we should think about how to make tests that would have detected any behavior change in the title and subtitle layout here.

nmoucht committed:
Handle <attachment> painting in GPUP
https://bugs.webkit.org/show_bug.cgi?id=246368
rdar://83459796

Reviewed by Simon Fraser.

Unify some parts of layout/painting of attachment on mac and ios.
This includes the struct containing layout info of the attachment
element, the functions constructing the lines of the subtitle and
title, painting of the lines. Move the bits constructing the
layout struct out of RenderThemeMac/iOS into AttachmentLayout.

* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::paint):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::AttachmentLayout::layOutSubtitle):
(WebCore::paintAttachmentIcon):
(WebCore::drawCTLine):
(WebCore::paintAttachmentTitle):
(WebCore::paintAttachmentSubtitle):

Canonical link: https://commits.webkit.org/255903@main
